[nnc] Make the pretty printer prettier

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57874 [nnc] Make the pretty printer prettier**

Before:
```
{
  for (int v = 0; v < 100; v++) {
    aten_sin[v] = sin(x_1[v]);
  }
{
    sum = float(0);
    for (int v_1 = 0; v_1 < 100; v_1++) {
      sum = ReduceOp((sum) + float(aten_sin[v_1]), reduce_args={});
    }
  }  for (int v_2 = 0; v_2 < 100; v_2++) {
    aten_cos[v_2] = cos(x_1[v_2]);
  }
  for (int v_3 = 0; v_3 < 100; v_3++) {
    aten_mul[v_3] = (_tensor_constant0[v_3]) * (aten_cos[v_3]);
  }
}
```

After:
```
{
  for (int v = 0; v < 100; v++) {
    aten_sin[v] = sin(x_1[v]);
  }
  {
    sum = float(0);
    for (int v_1 = 0; v_1 < 100; v_1++) {
      sum = ReduceOp((sum) + float(aten_sin[v_1]), reduce_args={});
    }
  }
  for (int v_2 = 0; v_2 < 100; v_2++) {
    aten_cos[v_2] = cos(x_1[v_2]);
  }
  for (int v_3 = 0; v_3 < 100; v_3++) {
    aten_mul[v_3] = (_tensor_constant0[v_3]) * (aten_cos[v_3]);
  }
}
```

Differential Revision: [D28455842](https://our.internmc.facebook.com/intern/diff/D28455842)

bertmaher committed:
[nnc] Make the pretty printer prettier

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -397,23 +397,24 @@ void IRPrinter::visit(const For* v) {
   indent_--;
   emitIndent();
-  os() << "}";
+  os() << "}\n";
```
Another fix could be to check if `s` is a `Block` on line 413 and emit indentation there - that would be a case when one block is nested directly into another and that's when we get those weird unindented braces like in your dump.

bertmaher replied:
Good point, I think there’s a better fix here than just suppressing the indent in for loops...

bertmaher committed:
Update on "[nnc] Make the pretty printer prettier"


Before:
```
{
  for (int v = 0; v < 100; v++) {
    aten_sin[v] = sin(x_1[v]);
  }
{
    sum = float(0);
    for (int v_1 = 0; v_1 < 100; v_1++) {
      sum = ReduceOp((sum) + float(aten_sin[v_1]), reduce_args={});
    }
  }  for (int v_2 = 0; v_2 < 100; v_2++) {
    aten_cos[v_2] = cos(x_1[v_2]);
  }
  for (int v_3 = 0; v_3 < 100; v_3++) {
    aten_mul[v_3] = (_tensor_constant0[v_3]) * (aten_cos[v_3]);
  }
}
```

After:
```
{
  for (int v = 0; v < 100; v++) {
    aten_sin[v] = sin(x_1[v]);
  }
  {
    sum = float(0);
    for (int v_1 = 0; v_1 < 100; v_1++) {
      sum = ReduceOp((sum) + float(aten_sin[v_1]), reduce_args={});
    }
  }
  for (int v_2 = 0; v_2 < 100; v_2++) {
    aten_cos[v_2] = cos(x_1[v_2]);
  }
  for (int v_3 = 0; v_3 < 100; v_3++) {
    aten_mul[v_3] = (_tensor_constant0[v_3]) * (aten_cos[v_3]);
  }
}
```


[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28455842).

bertmaher commented:
I think this version should leave things looking pretty nice.  Essentially I make `Block` responsible for all the whitespace, which I think works pretty well.

navahgar commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -401,14 +394,14 @@ void IRPrinter::visit(const For* v) {
   } else {
     os() << "{}";
   }
-  os() << std::endl;
```
Where are these printed now? I don't see them getting printed in `Block`.

bertmaher replied:
The newlines?  A newline is printed at the end of each statement in the body of a `Block`.

navahgar replied:
As I noted below, I don't see any newline at the end of each statement in a `Block`. Am I missing something?

Also, I see newlines at the end of some statements, like `Store`, `Let`, but not at the end of `For` and `Block`.

bertmaher replied:
Oh shoot.  I think I pushed an old version of this diff.  One moment.

navahgar commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -401,14 +394,14 @@ void IRPrinter::visit(const For* v) {
 
   for (Stmt* s : *v) {
+    emitIndent();
     os() << *s;
```
I don't see any newline here, after every statement in a `Block`.

bertmaher committed:
Update on "[nnc] Make the pretty printer prettier"


Before:
```
{
  for (int v = 0; v < 100; v++) {
    aten_sin[v] = sin(x_1[v]);
  }
{
    sum = float(0);
    for (int v_1 = 0; v_1 < 100; v_1++) {
      sum = ReduceOp((sum) + float(aten_sin[v_1]), reduce_args={});
    }
  }  for (int v_2 = 0; v_2 < 100; v_2++) {
    aten_cos[v_2] = cos(x_1[v_2]);
  }
  for (int v_3 = 0; v_3 < 100; v_3++) {
    aten_mul[v_3] = (_tensor_constant0[v_3]) * (aten_cos[v_3]);
  }
}
```

After:
```
{
  for (int v = 0; v < 100; v++) {
    aten_sin[v] = sin(x_1[v]);
  }
  {
    sum = float(0);
    for (int v_1 = 0; v_1 < 100; v_1++) {
      sum = ReduceOp((sum) + float(aten_sin[v_1]), reduce_args={});
    }
  }
  for (int v_2 = 0; v_2 < 100; v_2++) {
    aten_cos[v_2] = cos(x_1[v_2]);
  }
  for (int v_3 = 0; v_3 < 100; v_3++) {
    aten_mul[v_3] = (_tensor_constant0[v_3]) * (aten_cos[v_3]);
  }
}
```

Differential Revision: [D28455842](https://our.internmc.facebook.com/intern/diff/D28455842)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -428,43 +419,37 @@ void IRPrinter::visit(const Allocate* v) {
-    emitIndent();
     os() << "if (!" << *cond << ") ";
-    os() << *false_stmt << std::endl;
+    os() << *false_stmt;
```
I guess this would be a little yucky if we have a `Cond` with branches that aren't`Block`s.  That doesn't happen, empirically, but...

navahgar reviewed (commented):
Mostly LGTM. 

One minor issue with this approach is that the outermost statement will not have a newline at the end. For example, if the outer most statement is a `Block`, it will end without a newline. That is not ideal.

How about adding a newline in the method: `void IRPrinter::print(const Stmt& stmt)`? That seems to be the starting point for visiting all statements.

bertmaher commented:
> Mostly LGTM.
> 
> One minor issue with this approach is that the outermost statement will not have a newline at the end. For example, if the outer most statement is a `Block`, it will end without a newline. That is not ideal.
> 
> How about adding a newline in the method: `void IRPrinter::print(const Stmt& stmt)`? That seems to be the starting point for visiting all statements.

After I thought about it a bit, I didn't feel like it was a big deal to not end the outermost statement with a newline.  Setting aside the `Block` case I think I'd rather have no-newlines, since it's easier to do `std::cout << *stmt << "\n"` than to get rid of an embedded newline that I don't want.  `Block` is a bit weirder since I imagine you usually want `{\nstmts...\n}\n` but even there it's not hard to add `\n`, and things like GRAPH_DEBUG do it for you.

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28455842).
